### PR TITLE
Add support for cohere/titan embedding models

### PR DESCRIPTION
**Pull request by QuinnGT**
*Issue #, if available:*

Issues open that this resolves, #218 

*Description of changes:*

Brought over changes from spugachev 3 month old branch on cohere. Added changes for new schema. Deployed and tested in fresh environment last night but I see massi-ang opened #359 this morning. However cohere embedding needs additional changes for how it handles inputs. Documented [here](https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-embed.html).

search_document type and search_query type are used. Classification and clustering are not needed at this time. 

[embedding_types](https://docs.cohere.com/reference/embed) was not defined as float is appropriate for this implementation. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/shared/layers/python-sdk/python/genai_core/embeddings.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/embeddings.py:85]: 2 suggestions:
- Let's have providers in [types.py](https://github.com/aws-samples/aws-genai-llm-chatbot/blob/main/lib/shared/layers/python-sdk/python/genai_core/types.py) instead of hardcoded at each condition. 
- better to break down the provider-related code into separate:
  - `_generate_embeddings_amazon`
  - `_generate_embeddings_cohere` or any better naming idea 
  and just have the condition statement in this main `_generate_embeddings_bedrock`

File: lib/shared/layers/python-sdk/python/genai_core/embeddings.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/embeddings.py:104]: same for tasks let's type instead if hardcode them

**Review by bigadsoleiman — CHANGES_REQUESTED** [review]
Thanks for this submission!

Just added a couple of feedbacks

File: bin/config.ts
Comment (bigadsoleiman) [inline bin/config.ts:56]: please include [titan](https://github.com/aws-samples/aws-genai-llm-chatbot/pull/359/files#diff-45deac45daa48ef2d176c1d42cb9629f2607b020c8ea4bd8ab6763ac6033ac5fR37) from https://github.com/aws-samples/aws-genai-llm-chatbot/pull/359 which i closed in favor of this PR

**Review by bigadsoleiman — CHANGES_REQUESTED** [review]

File: bin/config.ts
Comment (QuinnGT) [reply bin/config.ts:56]: Added

File: lib/chatbot-api/functions/api-handler/routes/embeddings.py
Comment (massi-ang) [inline lib/chatbot-api/functions/api-handler/routes/embeddings.py:1]: let's use an Enum for task

**Review by massi-ang — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: bin/config.ts
Comment (QuinnGT) [reply bin/config.ts:56]: Actually one thought I had on this @bigadsoleiman is we are using Unstructured to to handle ingestion and partition all content, or at least I think it is. Would this conflict with Unstructured partition_image with ocr? I'll look into it and see if there is overlap on functions.

File: lib/shared/layers/python-sdk/python/genai_core/opensearch/query.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/opensearch/query.py:47]: use `Task.RETRIEVE.value`

File: lib/shared/layers/python-sdk/python/genai_core/aurora/query.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/aurora/query.py:48]: use `Task` enum

File: lib/shared/layers/python-sdk/python/genai_core/chunks.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/chunks.py:43]: use `Task` enum

File: lib/shared/layers/python-sdk/python/genai_core/workspaces.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/workspaces.py:115]: use `Task` enum

File: lib/shared/layers/python-sdk/python/genai_core/workspaces.py
Comment (bigadsoleiman) [inline lib/shared/layers/python-sdk/python/genai_core/workspaces.py:182]: use `Task` enum

**Review by bigadsoleiman — CHANGES_REQUESTED** [review]

**Comment by QuinnGT** [comment]
Quick update. I pushed updates for everything except for titan multimodal. Currently deploying those changes and testing before taking on titan multimodal. 

In regards to [Titan Multimodal Embeddings](https://docs.aws.amazon.com/bedrock/latest/userguide/titan-multiemb-models.html), it will need a bit more than just calling out the additional model. It will need support for it's API payload like max input text tokens and max input image size. However we are not capturing images in our use of Unstructured or rather lack of use of unstructured, like the `[partition_image](https://unstructured-io.github.io/unstructured/core/partition.html#partition-image)` function. Also I see unstructured is poorly used as we pull the full container yet only use the loader. 

As soon as the cohere changes finish and tests pass, I'll circle back on titan multimodal and unstructured. Feel free to post thoughts  or feedback.

**Comment by massi-ang** [comment]
`maxInputTokens` is optional and we currently only need to support text embeddings. This was already tested in #359 and working fine.

**Comment by QuinnGT** [comment]
Alright, all requested changes have been made. I also fixed the embedding prompt to show when enableRag was true vs just being skipped no matter what in the magic-config.

New environment has been deployed. Bedrock and Sagemaker embedding models tested as functional.

**Comment by QuinnGT** [comment]
Found issue with creating workspace. Not sure if it's from me or changes merged into this PR. Need to test and see.

File: cli/magic-config.ts
Comment (Rob-Powell) [inline cli/magic-config.ts:354]: this is wrong as if you select only kendra as your RAG you dont need an embedding model.

Comment (QuinnGT) [reply cli/magic-config.ts]: All good, fair point. Made a change so it only shows when it's aurora and/or opensearch.

**Comment by QuinnGT** [comment]
Resolved the latest issues. Feel free to test and let me know if anything else is needed.

File: lib/shared/layers/python-sdk/python/genai_core/embeddings.py
Comment (QuinnGT) [reply lib/shared/layers/python-sdk/python/genai_core/embeddings.py:114]: Sorry I missed that, good catch. I'll update it today.

**Comment by QuinnGT** [comment]
@bigadsoleiman I noticed an error in deployment that's happening for log groups due to size of name. I tested it  [here ](https://github.com/QuinnGT/aws-genai-llm-chatbot/commit/ba191410976277f13ea69aaddef463962d086964)and it works. Do you want me to bring that into this PR or wait till this is approved?

**Comment by massi-ang** [comment]
> @bigadsoleiman I noticed an error in deployment that's happening for log groups due to size of name. I tested it [here ](https://github.com/QuinnGT/aws-genai-llm-chatbot/commit/ba191410976277f13ea69aaddef463962d086964)and it works. Do you want me to bring that into this PR or wait till this is approved?

Hi Quinn, I checked your change and my concern is that if you deploy two stacks, they will share the same log group, since the value of `id` will be the same in for both.

**Comment by QuinnGT** [comment]
> > @bigadsoleiman I noticed an error in deployment that's happening for log groups due to size of name. I tested it [here ](https://github.com/QuinnGT/aws-genai-llm-chatbot/commit/ba191410976277f13ea69aaddef463962d086964)and it works. Do you want me to bring that into this PR or wait till this is approved?
> 
> Hi Quinn, I checked your change and my concern is that if you deploy two stacks, they will share the same log group, since the value of `id` will be the same in for both.

Great point @massi-ang . How about I use the config.prefix with the id so it looks something like `logGroupName: /aws/vendedlogs/states/${config.prefix}-${id}`. 

Before: /aws/vendedlogs/states/CreateAuroraWorkspace-CreateAuroraWorkspaceSMLogGroup
After: /aws/vendedlogs/states/dev-CreateAuroraWorkspaceSMLogGroup

**Comment by massi-ang** [comment]
I would suggest using `/aws/vendedlogs/states/${cdk.Stack.of(this).name}-${id}`, but there could still be issues in case we apply the same solution to other constructs which have the same `id` (since `id` in only unique in the scope of the parent construct).

**Comment by QuinnGT** [comment]
@bigadsoleiman is there anything else pending that I missed? If not, can you approve this PR? Thank you!

**Review by bigadsoleiman — APPROVED** [review]

**Comment by Cerrix** [comment]
Hi @QuinnGT,
I tested the pull request because I was looking for the Cohere Multilingual implementation. When I selected that model for embeddings, I received the following error response from AppSync:
`{
  "data": null,
  "errors": [
    {
      "path": [
        "calculateEmbeddings"
      ],
      "data": null,
      "errorType": "TypeError",
      "errorInfo": null,
      "locations": [
        {
          "line": 2,
          "column": 3,
          "sourceName": null
        }
      ],
      "message": "Object of type Task is not JSON serializable"
    }
  ]
}`
both creating the workspace or testing the embedding. 
With Titan everything is fine.

Thank you!

**Comment by QuinnGT** [comment]
> Hi @QuinnGT, I tested the pull request because I was looking for the Cohere Multilingual implementation. When I selected that model for embeddings, I received the following error response from AppSync: `{ "data": null, "errors": [ { "path": [ "calculateEmbeddings" ], "data": null, "errorType": "TypeError", "errorInfo": null, "locations": [ { "line": 2, "column": 3, "sourceName": null } ], "message": "Object of type Task is not JSON serializable" } ] }` both creating the workspace or testing the embedding. With Titan everything is fine.
> 
> Thank you!

@Cerrix  thank you for reporting this. Can you share what RAG you were testing with?

**Comment by Cerrix** [comment]
Yep: I’m testing in with Open Search (and just that). I get the same error even if testing cohere through the “embeddings” page (not only creating a new workspace).

thank you

**Comment by QuinnGT** [comment]
> Yep: I’m testing in with Open Search (and just that). I get the same error even if testing cohere through the “embeddings” page (not only creating a new workspace).
> 
> thank you

Hey @Cerrix I enabled and deployed an opensearch instance and I'm not seeing any errors with workspace creation or testing embeddings. I'm not sure if it's a new or existing deployment you are testing in but try doing a fresh deployment from the latest, along with a new npm install and build. Feel free to report back with more information to troubleshoot, if needed.

**Comment by QuinnGT** [comment]
Comparison of embedding models from demo data. 
image

**Comment by Cerrix** [comment]
It was a week old deploy. But let me try again from scratch! I'll keep you posted!

**Comment by Cerrix** [comment]
It is so strange. I re-deployed from scratch the repository branching your PR in Oregon (us-west-2).

Here is my config file: 
```
{
  "prefix": "llmchatbot-demo",
  "privateWebsite": false,
  "certificate": "",
  "domain": "",
  "bedrock": {
    "enabled": true,
    "region": "us-west-2"
  },
  "llms": {
    "sagemaker": []
  },
  "rag": {
    "enabled": true,
    "engines": {
      "aurora": {
        "enabled": false
      },
      "opensearch": {
        "enabled": true
      },
      "kendra": {
        "enabled": false,
        "createIndex": false,
        "external": [],
        "enterprise": false
      }
    },
    "embeddingsModels": [
      {
        "provider": "sagemaker",
        "name": "intfloat/multilingual-e5-large",
        "dimensions": 1024
      },
      {
        "provider": "sagemaker",
        "name": "sentence-transformers/all-MiniLM-L6-v2",
        "dimensions": 384
      },
      {
        "provider": "bedrock",
        "name": "amazon.titan-embed-text-v1",
        "dimensions": 1536
      },
      {
        "provider": "bedrock",
        "name": "amazon.titan-embed-image-v1",
        "dimensions": 1024
      },
      {
        "provider": "bedrock",
        "name": "cohere.embed-english-v3",
        "dimensions": 1024
      },
      {
        "provider": "bedrock",
        "name": "cohere.embed-multilingual-v3",
        "dimensions": 1024,
        "default": true
      },
      {
        "provider": "openai",
        "name": "text-embedding-ada-002",
        "dimensions": 1536
      }
    ],
    "crossEncoderModels": [
      {
        "provider": "sagemaker",
        "name": "cross-encoder/ms-marco-MiniLM-L-12-v2",
        "default": true
      }
    ]
  }
}
```
Still I got the same error:
Screenshot 2024-03-04 at 00 06 01

**Comment by QuinnGT** [comment]
@Cerrix  Interesting, I'm not sure. It looks like it's an error from the appsync function. Almost like the old version of app sync functions are trying to be used with the newer genai_core python functions. Did you purge your cdk.out when doing a fresh deployment? Also double check cdk version is updated on your local as there seems to be some rebuild bug introduced on each release.

@massi-ang I see Bigad approved the PR but we're still waiting on you or another repo owner for approval. What else do you need to from me?

**Comment by Cerrix** [comment]
I'm deploying following the Cloud9 instructions of the sample documentation: https://aws-samples.github.io/aws-genai-llm-chatbot/guide/deploy.html

If we need to update some dependencies I think it must be documented before merging the PR in main.

I can try with a vanilla cloud9 instance but we need to double check the issue. I don't know if someone from the team would like to try the deploy!

File: lib/shared/layers/python-sdk/python/genai_core/types.py
Comment (massi-ang) [inline lib/shared/layers/python-sdk/python/genai_core/types.py:35]: This is confusing semantic as Bedrock and Sagemaker are not Provider in the same sense as Amazon and Cohere are. Please split Model Provider from Hosting Providers.

**Review by massi-ang — CHANGES_REQUESTED** [review]
Please fix the `Provider` type with the right semantics

File: lib/shared/layers/python-sdk/python/genai_core/types.py
Comment (QuinnGT) [reply lib/shared/layers/python-sdk/python/genai_core/types.py:35]: I agree, it's not clean. I'll split it and push today.

Comment (QuinnGT) [reply lib/shared/layers/python-sdk/python/genai_core/types.py:35]: Splitting the providers into `ModelProviders` and `HostingProviders` has a way bigger impact than I expected. `Provider `is widely used and `ModelProvider` is oddly used in the react-app/common/types. At this point I've made changes in 20 files but the sprawl of changes needed continues. 

Do you want me to continue to do the split or am I just approaching this incorrectly and overcomplicating it?

Comment (bigadsoleiman) [reply lib/shared/layers/python-sdk/python/genai_core/types.py:35]: @QuinnGT while I agree there should be a separation the give the ramification of this split I'd say is out of the scope of this PR. So if you want to revert back to just provider we can create a separate issue for this refactor.

Comment (QuinnGT) [reply lib/shared/layers/python-sdk/python/genai_core/types.py:35]: @bigadsoleiman okay, I didn't commit my changes so it should all be stable as is now. Fork is sync'd with main now. 

You can approve the PR and close it.

**Review by bigadsoleiman — APPROVED** [review]

**Pull request merged**
